source/common/filter/tcp_proxy.cc

```diff
@@ -113,13 +127,23 @@ TcpProxy::~TcpProxy() {
 
   if (upstream_connection_) {
     finalizeUpstreamConnectionStats();
+    if (upstream_connection_->state() == Network::Connection::State::Open) {
```

Comment:
I think this logic should be done in the context of `onDownstreamEvent`. Any reason not to do it there? If so, the stats finalizing might need to get moved there also.

Reply:
I agree.  This also made me realize that we can't finalize the upstream connections until the connection is closed by the drainer.  So I'm fixing that as well.